### DISCUSSION TITLE
Can we update table item and remove secondary index using same query?

### DISCUSSION THREAD

**Discussion by hemanth1226** [sensedeep/dynamodb-onetable#55] [Q&A] [answered]
I am trying to handle a scenario where we have to update a table item with new value as well as remove the secondary index (gsi). Is there a way we can handle that? 

Please find the example below.

``` javascript
schema = {
    indexes: {
      primary: { hash: 'pk', sort: 'sk' },
      gsi1: { hash: 'gsi1pk', sort: 'gsi1sk', project: 'all', follow: true },
      gsi2: { hash: 'gsi2pk', sort: 'gsi2sk', follow: true },
      gsi3: { hash: 'gsi3pk', sort: 'gsi3sk', follow: true },
    },
    models: {
      Mailbox: {
        pk: { type: String, value: 'mail#${id}' },
        sk: { type: String, value: 'mail#${id}' },
        read: { type: Boolean },
        subject: { type: String },
        desc: { type: String },
        body: { type: String },
        userId: { type: String, required: true },
        id: { type: String, required: true, uuid: 'ulid' },
        gsi1pk: {
          type: String,
          value: 'user#${userId}',
        },
        gsi1sk: {
          type: String,
          value: 'mail#${id}',
        },
        gsi2pk: {
          type: String,
          value: 'user#${userId}#unread',
        },
        gsi2sk: {
          type: String,
          value: 'mail#${id}',
        },
        _type: { type: String, filter: false, value: 'Mailbox' },
      },
    },
  };
```

``` javascript
table = new Table({
    client: client,
    name: 'table-name',
    delimiter: '#',
    schema: schema,
  });
```

```
  mailboxModel = this.table.getModel('Mailbox');
```

Here I need the GSI 2 to only include the mails that are unread.

When i try the following, it's just removing GSI2 but not updating the `read` status in main table. 

``` javascript
return this.mailBoxModel.update(data, {
      log: true,
      updateIndexes: true,
      remove: data.read ? ['gsi2pk', 'gsi2sk'] : [],
    });
```

Can you please check and let me know if there is a better way to handle this scenario?

Appreciate your help. Thanks!

**Comment by mobsense** [comment]
Sure, can you please post the log of the actual command generated (via log: true)?

**Comment by mobsense** [comment, marked as answer]
I can confirm this is a bug.  Currently, the remove[] overrides any updates.

We've got a fix in 1.4.8 (just committed) and will continue to test and then update NPM. 

Your welcome to check the Git repo an see if the fix works for you.

**Reply by hemanth1226** [reply]
Thank you so much. It works perfectly.

**Reply by mobsense** [reply]
No worries. Thanks for reporting.

**Reply by mobsense** [reply]
Just a heads up. Further testing has revealed that this fix isn't quite right when handling add/delete/remove/set operations in various combinations. Will patch today.

Also, this update provides a much easier way to remove attributes .... set to null. i.e.

If a value is undefined, it isn't updated. If null, the attribute is removed.

You can modify this behavior by setting the Table params.nulls to true which permits writing nulls to the database. This can also be modified on a per schema model basis. The default is params.nulls is false ... i.e. do not write nulls to the database and instead remove attributes set to null. NOTE: this only impacts the item being updated.

```
data.gsi2pk = null
data.gsi2sk = null
mailBoxModel.update(data)
```
